src/node/media_stream_destination.rs

```diff
@@ -0,0 +1,163 @@
+/// let osc = context.create_oscillator();
+///
+/// // Create a media destination node
+/// let dest = context.create_media_stream_destination();
+/// osc.connect(&dest);
+/// osc.start();
+///
+/// // Handle recorded buffers
+/// println!("samples recorded:");
+/// let mut samples_recorded = 0;
+/// for item in dest.stream() {
+///     let buffer = item.unwrap();
+///
+///     // You could write the samples to a file here.
+///     samples_recorded += buffer.length();
+///     print!("{}\r", samples_recorded);
+/// }
+/// ```
+///
+/// # Examples
+///
+/// - `cargo run --release --example recorder`
+
+pub struct MediaStreamAudioDestinationNode {
+    registration: AudioContextRegistration,
+    channel_config: ChannelConfig,
+    receiver: Receiver<AudioBuffer>,
+}
+
+impl AudioNode for MediaStreamAudioDestinationNode {
+    fn registration(&self) -> &AudioContextRegistration {
+        &self.registration
+    }
+
+    fn channel_config_raw(&self) -> &ChannelConfig {
+        &self.channel_config
+    }
+
+    fn number_of_inputs(&self) -> u32 {
+        1
+    }
+
+    fn number_of_outputs(&self) -> u32 {
+        0
+    }
+}
+
+impl MediaStreamAudioDestinationNode {
+    /// Create a new MediaStreamAudioDestinationNode
+    pub fn new<C: BaseAudioContext>(context: &C, options: ChannelConfigOptions) -> Self {
+        context.base().register(move |registration| {
+            let (send, recv) = crossbeam_channel::bounded(1);
+            let recv_control = recv.clone();
+
+            let node = MediaStreamAudioDestinationNode {
+                registration,
+                channel_config: options.into(),
+                receiver: recv_control,
+            };
+
+            let render = DestinationRenderer { send, recv };
+
+            (node, Box::new(render))
+        })
+    }
+
+    /// A [`MediaStream`] iterator producing audio buffers with the same number of channels as the
+    /// node itself
+    ///
+    /// While you could call this function multiple times and poll all iterators concurrently, each
+    /// buffer will only be offered once.
+    pub fn stream(&self) -> impl MediaStream {
+        AudioDestinationNodeStream {
+            receiver: self.receiver.clone(),
+        }
+    }
+}
+
+struct DestinationRenderer {
+    send: Sender<AudioBuffer>,
+    recv: Receiver<AudioBuffer>,
+}
+
+impl AudioProcessor for DestinationRenderer {
+    fn process(
+        &mut self,
+        inputs: &[AudioRenderQuantum],
+        _outputs: &mut [AudioRenderQuantum],
+        _params: AudioParamValues,
+        _timestamp: f64,
+        sample_rate: SampleRate,
+    ) -> bool {
+        // single input, no output
+        let input = &inputs[0];
+
+        // convert AudioRenderQuantum to AudioBuffer
+        let samples: Vec<_> = input
+            .channels()
+            .iter()
+            .map(|c| c.as_slice().to_vec())
+            .collect();
+        let buffer = AudioBuffer::from(samples, sample_rate);
+
```

Comment:
This construction with a sender and a receiver at the same time is a bit of a hack, but it works.
We could make it a bit more lightweight by utilizing an `Arc<Option<AudioBuffer>>` guarded by a `CondVar`

Reply:
But is this needed as the channel is bounded to `1`? I was thinking that it meant that the channel could only hold 1 message at a time, automatically discarding the previous one if it has not been consumed

Reply:
Nope it works differently. Check the first example at https://docs.rs/crossbeam-channel/latest/crossbeam_channel/fn.bounded.html
When the channel is full, the sender will block until there is capacity again

Reply:
Oh ok I see, thanks